OAK-9131 : introducing scanners for references to nt:frozenNode

stefan-egli opened the pull request.

stefan-egli committed:
OAK-9131 : introducing scanners for references to nt:frozenNode

mreutegg commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/FrozenNodeRefsByScanningCommand.java:
```diff
@@ -0,0 +1,293 @@
+public class FrozenNodeRefsByScanningCommand implements Command {
+
+    static {
+        // disable any query limits as our query is going to be a fulls can, and we are aware of it
```
```suggestion
        // disable any query limits as our query is going to be a full scan, and we are aware of it
```

stefan-egli replied:
thx! fixed

mreutegg commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/FrozenNodeRefsByScanningCommand.java:
```diff
@@ -0,0 +1,293 @@
+
+    public static final String NAME = "frozennoderefsbyscanning";
+
+    private final String summary = "Scans and lists all references to nt:frozenNode";
```
```suggestion
    private final String summary = "Scans repository and lists all references to nt:frozenNode";
```

stefan-egli replied:
thx! fixed

mreutegg commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/FrozenNodeRefsByScanningCommand.java:
```diff
@@ -0,0 +1,293 @@
+            System.out.println("Logging in...");
+            Session session = openSession(nodeStore, "crx.default", user, password);
+
+            if (createRefPath != null && createRefProp != null && createRefUuid != null) {
```
I would move this part into a dedicated test class.

stefan-egli replied:
Agree, much nicer. Done now at https://github.com/apache/jackrabbit-oak/pull/238/commits/fb8b330c4f248078bb2dce62cfb19ae90a50030c

mreutegg commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/FrozenNodeRefsByScanningCommand.java:
```diff
@@ -0,0 +1,293 @@
+            String referrerPath = n.getPath();
+            String referrerProperty = p.getName();
+            FrozenNodeRef ref = new FrozenNodeRef(referrerPath, referrerProperty, PropertyType.nameFromValue(p.getType()), uuid, path);
+            list.add(ref);
```
Won't this consume too much memory when there are many references? The only use of the list seems to be to decide on the return of `uuidscan()`.

stefan-egli replied:
Good point. The list was more for reporting each find - but ultimately there's no other use case for this atm, and the reporting itself can done as it happens. So I've now changed this to only carry a `count` at https://github.com/apache/jackrabbit-oak/pull/238/commits/fb8b330c4f248078bb2dce62cfb19ae90a50030c

mreutegg commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/FrozenNodeRefsUsingIndexCommand.java:
```diff
@@ -0,0 +1,159 @@
+            System.out.println("Scanning ... refindex = " + refIndex);
+            NodeState references = refIndex.getChildNode(":references");
+
+            List<FrozenNodeRef> list = new LinkedList<FrozenNodeRef>();
```
Same concern about memory as above.

stefan-egli replied:
Agree - changed to a `count` at https://github.com/apache/jackrabbit-oak/pull/238/commits/fb8b330c4f248078bb2dce62cfb19ae90a50030c

mreutegg commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/FrozenNodeRefsUsingIndexCommand.java:
```diff
@@ -0,0 +1,159 @@
+            NodeStore store = Utils.bootstrapNodeStore(nopts, closer);
+            NodeState root = store.getRoot();
+
+            NodeBuilder b = root.builder();
```
Move to test class?

stefan-egli replied:
Oups, this one is particularly ugly - removed this part entirely (at https://github.com/apache/jackrabbit-oak/pull/238/commits/fb8b330c4f248078bb2dce62cfb19ae90a50030c )

mreutegg reviewed (commented):
In general, can those two commands run in read-only mode? This would make them safe to run on data that is in use. E.g. a production system.

Could you please also update the README.md?

stefan-egli committed:
OAK-9131 : incorporated review feedback

stefan-egli commented:
@mreutegg thx for the review - I've done the changes in [this commit](https://github.com/apache/jackrabbit-oak/pull/238/commits/fb8b330c4f248078bb2dce62cfb19ae90a50030c)

> In general, can those two commands run in read-only mode? This would make them safe to run on data that is in use. E.g. a production system.

Yes, and with the suggested move of the test code to a test class this is now also more clear: only the test class uses the `-read-write=true` parameter - so even more reason to move test code to a test class indeed.
> Could you please also update the README.md?

Yes, will add next, thx for the pointer

stefan-egli committed:
OAK-9131 : frozennoderefsbyscanning and frozennoderefsusingindex documented

stefan-egli commented:
> > Could you please also update the README.md?
>
> Yes, will add next, thx for the pointer

Now done in [this commit](https://github.com/apache/jackrabbit-oak/pull/238/commits/7a0e090c200b4081416393d0a8bcbb643dd87945)

mreutegg commented on oak-run/README.md:
```diff
@@ -136,6 +138,42 @@ browsing of an existing oak repository.
+frozennoderefsbyscanning
+------------------------
+
+This command executes a potentially expensive (!) travering query searching for
```
```suggestion
This command executes a potentially expensive (!) traversing query searching for
```

stefan-egli replied:
committed, thx

mreutegg commented:
> > In general, can those two commands run in read-only mode? This would make them safe to run on data that is in use. E.g. a production system.
> 
> Yes, and with the suggested move of the test code to a test class this is now also more clear: only the test class uses the `-read-write=true` parameter - so even more reason to move test code to a test class indeed.

Hmm, it still looks like the scan command starts a repository in read-write mode.

stefan-egli committed:
OAK-9131 : typo

Co-authored-by: Marcel Reutegger <marcel.reutegger@gmail.com>

stefan-egli committed:
OAK-9131 : Adding readOnlyMode to Utils.bootstrapNodeStore and with that explicitly setting readOnly mode for both frozenNode scanner variants

stefan-egli committed:
OAK-9131 : minor sysout

stefan-egli commented:
> Hmm, it still looks like the scan command starts a repository in read-write mode.

Indeed, thx for the heads-up!

Fixed it now:
* the `frozennoderefsbyscanning` variant already opened in readOnly mode, unless `-read-write=true` was set. Now explicitly disabled that option and therefore always using readOnly.
* the `frozennoderefsusingindex` variant was indeed opening in readWrite (!) mode. As it is using a completely different way of accessing the NodeStore, that utils variant didn't yet support readOnly mode - so I've now added this, and used it for this case.

stefan-egli commented:
Merged to trunk in rev [1879938](http://svn.apache.org/viewvc?rev=1879938&view=rev)
